drivers: ethernet: Add Intel® i226 Ethernet Controller Driver

Vijayakannan opened the pull request:
The Intel i226 Ethernet Controller is a PCIe Gen 2 one-lane modular endpoint device that integrates a fully featured GbE Media Access Control (MAC) and Physical Layer (PHY) port. This PR introduces a complete driver bundle to enable Intel i226 Ethernet Controller.

**Components:**
- MAC Driver, MDIO Driver & Parent Platform Driver

**Features Supported:**
- Full duplex support for 10/100/1000 Mbps.
- Half duplex support for 10/100 Mbps.
- MSI-X interrupt support for TX/RX DMA channels.
- Exclusive bottom-half processing for each hardware queue.
- Auto-Negotiation for 10/100/1000 Mbps.
- MTU customization for flexible packet sizes.
- MAC address filtering support for:
     - Random MAC Address Generated.
     - MAC addresses specified in the device tree.
     - Pre-programmed MAC addresses in EEPROM.
- MDIO Clause 22 (C22) and Clause 45 (C45) support.
- Multi-instance support to scale the system. (Tested Up to 2 instances with basic coverage)

**Other Notes:**
- The Parent Platform Driver is designed to be modular and reusable for other PCIe-based Ethernet devices.
- 2.5 Gbps, Time Sync and other TSN features were not enabled despite the hardware having support.

 **Validation Coverage:**
This driver bundle has been tested on Intel Alder Lake platforms with the Intel i226 LM variant Ethernet Controller(0x125B). The ZPERF network sample application was used to validate basic functionality, performance, and stability.

Test List:
- Transmit and receive UDP packet,
- Transmit and receive TCP packet,
- Speed: 10/100/1000Mbps, Duplex: Full, Auto-negotiation: ON, (autoneg initiated from link partner),
- Speed: 10/100; Duplex: Half, Auto-negotiation: ON, (autoneg initiated from link partner),
- Speed: 10/100/1000Mbps, Duplex: Full, Auto-negotiation: ON (autoneg initiated from net shell),
- Speed: 10/100Mbps, Duplex: Half, Auto-negotiation: ON (autoneg initiated from net shell),
- TX and RX with different MTU sizes: 256, 512, and 1024
- Throughput KPI for UDP upload and download (100/1000 Mbps),
- Latency KPI: 100 Mbps, 500 us packet interval with 500,1500 bytes packet size,
- Latency KPI: 1000 Mbps, 125 us packet interval, 200 bytes packet size,
- Ethernet Cable hot-plug with different number of hardware queues.
- MDIO Clause - 22/ 45 Support,
- Configure MAC address via net shell, device tree and net shell.

maass-hamburg commented:
On a first look:

- the Link status change detection should happen in the phy. we have the ``phy_link_callback_set()`` to set a callback function in the phy from the mac. In that callback ``net_eth_carrier_on()`` and ``net_eth_carrier_off()`` should be called.
- It's generall best to not have code for the phy in the ethernet driver. It's preferd if the official zephyr apis are used for communication between the phy and mac driver.

Vijayakannan commented:
> On a first look:
> 

Hi Fin Maaß, Thanks for spending time to review and recommendations.

> * the Link status change detection should happen in the phy. we have the `phy_link_callback_set()` to set a callback function in the phy from the mac. In that callback `net_eth_carrier_on()` and `net_eth_carrier_off()` should be called.
> * It's generall best to not have code for the phy in the ethernet driver. It's preferd if the official zephyr apis are used for communication between the phy and mac driver.

I agree with your suggestions and work on it.

**Rationale on why earlier decided to use MAC for Link detection:**
Intel i226 integrates the GbE MAC and PHY into a single chip. Additionally, MAC includes automatic Link detection functionality and update its registers automatically, which simplifies usage. So, in my initial design opted to handle it directly in MAC to avoid the overhead of invoking PHY-to-MAC callback. 

I am open for your suggestion in case if you have other thoughts. - Thanks, Vijay

Vijayakannan commented on drivers/mdio/mdio_intel_igc.c:
```diff
@@ -0,0 +1,200 @@
+	const struct intel_igc_mdio_cfg *cfg = dev->config;
+	struct intel_igc_mdio_data *data = dev->data;
+
+	data->base = DEVICE_MMIO_GET(cfg->platform);
```
> maybe it is possible to save the base in ` struct intel_igc_mdio_cfg` and already assign it in `INTEL_IGC_MDIO_INIT()` ?

DEVICE_MMIO_GET will not produce the base value during compile time. Thus, try saving it in the INTEL_IGC_MDIO_INIT() macro is actually leading to build error.

Vijayakannan commented:
> > On a first look:
> 
> Hi Fin Maaß, Thanks for spending time to review and recommendations.
> 
> > * the Link status change detection should happen in the phy. we have the `phy_link_callback_set()` to set a callback function in the phy from the mac. In that callback `net_eth_carrier_on()` and `net_eth_carrier_off()` should be called.
> > * It's generall best to not have code for the phy in the ethernet driver. It's preferd if the official zephyr apis are used for communication between the phy and mac driver.
> 
> I agree with your suggestions and work on it.
> 

Progress Update:
The interrupt line used for this PHY link status change is shared for multiple purposes as per the hardware design. Therefore, I have started a small refactor to accommodate this review comment and all other use cases. Let me publish the changes once I finish my sanity checks.

Vijayakannan commented on drivers/ethernet/phy/phy_intel_igc.c:
```diff
@@ -0,0 +1,345 @@
+	} else {
+		mii_1ktcr_val &= ~MII_ADVERTISE_1000_FULL;
+	}
+
```
> There is a `LINK_FULL_2500BASE_T`, that could be implemented here too, as the phy supports it.

Let me follow up with another PR on this 2500BASE_T support.

Vijayakannan commented:
> > > On a first look:
> > 
> > 
> > Hi Fin Maaß, Thanks for spending time to review and recommendations.
> > > * the Link status change detection should happen in the phy. we have the `phy_link_callback_set()` to set a callback function in the phy from the mac. In that callback `net_eth_carrier_on()` and `net_eth_carrier_off()` should be called.
> > > * It's generall best to not have code for the phy in the ethernet driver. It's preferd if the official zephyr apis are used for communication between the phy and mac driver.
> > 
> > 
> > I agree with your suggestions and work on it.
> 
> Progress Update: The interrupt line used for this PHY link status change is shared for multiple purposes as per the hardware design. Therefore, I have started a small refactor to accommodate this review comment and all other use cases. Let me publish the changes once I finish my sanity checks.

Please review the updated driver version and share your feedback. Thanks in advance.

**Brief about this version: **
The PHY link status change is part of the Miscellaneous MSI-X interrupt vector. This vector is shared among PHY link status changes, time synchronization, RX write-back, RX miss, TX write-back, and other events. With this updated driver version, the functionality now extends beyond handling PHY link state changes.

Additionally, apart from decoupling the MISC interrupt handling, I have made significant refinements in this new version.

pdgendt commented on drivers/mdio/mdio_intel_igc.c:
```diff
@@ -0,0 +1,200 @@
+	ARG_UNUSED(dev);
+}
+
+static const struct mdio_driver_api mdio_api = {
```
```suggestion
static DEVICE_API(mdio, mdio_api) = {
```

pdgendt commented on drivers/ethernet/intel/CMakeLists.txt:
```diff
@@ -0,0 +1,5 @@
+# Copyright (c) 2025 Intel Corporation.
+# SPDX-License-Identifier: Apache-2.0
+
+zephyr_library_sources_ifdef(CONFIG_ETH_INTEL_PLAT	eth_intel_plat.c)
+zephyr_library_sources_ifdef(CONFIG_ETH_INTEL_IGC	eth_intel_igc.c)
```
Mixing tabs and spaces? Also prefer to sort alphabetically

```suggestion
zephyr_library_sources_ifdef(CONFIG_ETH_INTEL_IGC	eth_intel_igc.c)
zephyr_library_sources_ifdef(CONFIG_ETH_INTEL_PLAT eth_intel_plat.c)
```

Vijayakannan replied:
> Mixing tabs and spaces? Also prefer to sort alphabetically

Resolved in the new version. Please check.

pdgendt commented:
Please take a look at the issues sonarqubecloud reports.

Vijayakannan commented:
Hi @maass-hamburg and All reviewers,

Please review the updated PR version and feedback your recommendation. Thanks in advance.

changes since v2:
- rework the mac_address usage as per suggestion from @maass-hamburg 
	1) zephyr,random-mac-address
	2) local-mac-address
	3) something device specific, like here the eeprom or using the hwinfo
- rewrote mac_addr usage code to increase readability.
- remove intel_igc_mdio_bus_enable and intel_igc_mdio_bus_disable 
- move base getter to specific call - intel_igc_mdio
- remove multi-gigabit dts, yaml and respective phy code paths.
- remove mac_addr_exist_in_nvm dts, yaml and code paths.
- align macros end of line with '\' in all 4 drivers.
- fix macro alignments in the eth_intel_igc_priv.h header file.
- move SKU validation checker(eth_intel_validate_sku) to eth_intel_plat.c
- fix sonarqubecloud bot reported minor issues.

maass-hamburg commented on drivers/ethernet/phy/phy_intel_igc.c:
```diff
@@ -0,0 +1,371 @@
+		LOG_ERR("Failed to restart Autoneg");
+	}
+
+	k_work_init_delayable(&data->link_work, update_link_state);
```
this should be init before the interrupt is activated. it would also be good to schedule a run at the end of this function, so we get the current state, as the phy could already be up. maybe use `Z_WORK_DELAYABLE_INITIALIZER` in `INTEL_IGC_PHY_INIT`

Vijayakannan commented on drivers/ethernet/phy/phy_intel_igc.c:
```diff
@@ -0,0 +1,371 @@
+	const struct device *dev = data->phy;
+	struct phy_link_state new_state = {0};
+
+	if (phy_intel_igc_get_link_state(dev, &new_state) < 0) {
+		LOG_ERR("PHY Link is Down");
+	}
```
> it is not an error, if the link is down, at least not in the phy, maximum it's a info.
> 

Ok. I will change it as LOG_INF.

> the function `phy_intel_igc_await_phy_status()` in there does register reads until a specific value is there (blocking). we don't want that here, we want the current state here and don't want to wait for change, thats what we have the interrupt for. Just read the register once.

Sure. Let me cross check on it work after removing this blocking read.

BTW, update_link_state runs in the bottom-half processing. May I know why zephyr don't allow blocking call?

maass-hamburg replied:
blocking call is fine here in general. it was function wise. `phy_intel_igc_await_phy_status()` seems like it waits for autoneg beeing finished. I unfortunately didn't found the datasheet or manual which said, when the phy interrupt is triggered, but i supposed if it really is a link up or down trigger, then checking for finished auto-negation is not necessary. By the way if you have such documentation with the registers and so on please upload here or send me f.maass@vogl-electronic.com

Vijayakannan replied:
> blocking call is fine here in general. it was function wise. phy_intel_igc_await_phy_status() seems like it waits for autoneg beeing finished. I unfortunately didn't found the datasheet or manual which said, when the phy interrupt is triggered, but i supposed if it really is a link up or down trigger, then checking for finished auto-negation is not necessary. 

According to the specification, all PHY events are signaled as a single aggregated interrupt. The MII_BMSR_LINK_STATUS bit is primarily used to check the link status during any such event. I will update the phy_intel_igc_await_phy_status() function to clarify this and avoid any misconceptions.

> By the way if you have such documentation with the registers and so on please upload here or send me [f.maass@vogl-electronic.com](mailto:f.maass@vogl-electronic.com)

TODO: I am waiting for internal response, and I will keep you updated.

Vijayakannan commented:
Hi All,

Please review the updated PR version 5 and share your feedback. Thanks in advance, Vijay

**Changes since v4:**
	- Remove net_if_is_carrier_ok.
	- Combine link_state and link_speed getters as one function phy_intel_igc_get_link_state.
	- Get link status directly using phy_intel_igc_read.
	- Use k_work_reschedule for differed interrupt context.
	- Move differed work_init to driver init macro - INTEL_IGC_PHY_INIT.
        - Add eth_intel_igc_get_phy for get_phy 
	- Tune the LOG_INF/ERR prints.
	- Removed ETHERNET_DUPLEX_SET and ETHERNET_CONFIG_TYPE_LINK as per latest kernel version.
	- Add explicit condition check.
	- Keep ret for each usage of phy_intel_igc_read.

Vijayakannan commented on drivers/ethernet/phy/phy_intel_igc.c:
```diff
@@ -0,0 +1,383 @@
+		return ret;
+	}
+
+	ret = phy_intel_igc_aneg_10_100mbps(dev);
+	if (ret < 0) {
+		LOG_ERR("Autoneg init failed for 10/100 Mbps");
+		return ret;
+	}
+
+	ret = phy_intel_igc_aneg_1000mbps(dev);
+	if (ret < 0) {
+		LOG_ERR("Autoneg init failed for 1000 Mbps");
+		return ret;
+	}
```
> just use phy_intel_igc_cfg_link() instead with the use of the new dt prop (#89210)

Ok. I will update according to #89210 and #91572. Are there any other comments to accommodate in the next version of this PR?

Vijayakannan commented:
Hi @maass-hamburg, please approve if there are no more comments from your side.

Vijayakannan commented:
> Hi @maass-hamburg, please approve if there are no more comments from your side.

Hi @maass-hamburg, just a reminder to add this PR in your review list. Thanks in advance.

maass-hamburg commented:
@Vijayakannan I'm still not happy about the phy irq in the eth driver and not being in the platform driver.
But we could shorten it here and you can just for now use the generic `ethernet-phy` driver and add your specific phy driver with the correct irq handling later.

Vijayakannan commented:
> @Vijayakannan I'm still not happy about the phy irq in the eth driver and not being in the platform driver. But we could shorten it here and you can just for now use the generic `ethernet-phy` driver and add your specific phy driver with the correct irq handling later.

sure. I have replied to the earlier review comment to keep track of this discussion.

maass-hamburg commented on drivers/ethernet/intel/eth_intel_igc.c:
```diff
@@ -0,0 +1,1432 @@
+#define NUM_RX_DESC(n) DT_INST_PROP(n, num_rx_desc)
+#define NUM_MSIX(n)    NUM_QUEUES(n) + ETH_IGC_NUM_MISC
+
+#define INTEL_IGC_MAC_ASSIGN_ADDR(n)                                                               \
+	static void assign_mac_address_##n(uint8_t mac_addr[INTEL_IGC_ALEN])                       \
+	{                                                                                          \
+		const uint8_t mac_addr_dt[INTEL_IGC_ALEN] = DT_INST_PROP(n, local_mac_address);    \
+		memcpy(mac_addr, mac_addr_dt, sizeof(mac_addr_dt));                                \
+	}
```
seems unused

Vijayakannan replied:
> seems unused

Yes. I will clean up in the next version.

Vijayakannan committed:
drivers: ethernet: Add platform driver for MMIO mapping reuse

The Ethernet device model consists of multiple subsystem components, such
as MDIO, PHY, MAC and PTP_CLOCK. These components are mapped into a single
PCIe BAR location with same base address.

This platform driver retrieves the MMIO mapping details and provides a
framework to share it with all the child subsystem components. This
approach avoid the duplicate remapping, ensures efficient re-use of
MMIO mappings across related devices.

Example device tree structure for first ethernet instance:

parent0: parent0 {
        compatible = "intel,eth-plat";
        interrupt-parent = <&intc>;
        vendor-id  = <0x8086>;
        device-id  = <0xXXXX>;

        igc0: igc0 {
                compatible = "intel,igc-mac";

                /*
                 * MAC specific properties.
                 */

                status = "okay";
        };

        mdio0: mdio0 {
                compatible = "intel,igc-mdio";
                #address-cells = <1>;
                #size-cells = <0>;

                ethphy0: ethernet-phy@0 {
                        compatible = "ethernet-phy";
                        /*
                         * PHY specific properties.
                         */
                        reg = <0x0>;
                };
        };
};

This framework is modular and re-usable for other PCIe based Ethernet
devices. It can also be extended to support additional platform specific
information shared across child nodes.

Signed-off-by: Vijayakannan Ayyathurai <vijayakannan.ayyathurai@intel.com>

Vijayakannan committed:
drivers: mdio: Add Intel i226 MDIO driver support

Intel i226 MAC supports MDIO C22 and MDIO C45. Standard PHY registers
are accessible through MDIO C22, whereas PMAPMD and PCS are accssible
through MDIO C45.

Signed-off-by: Vijayakannan Ayyathurai <vijayakannan.ayyathurai@intel.com>

Vijayakannan committed:
drivers: ethernet: intel: Add i226 Ethernet MAC device driver

The Intel i226 Ethernet Controller is a PCIe Gen 2 one-lane modular
endpoint device that integrates a GbE Media Access Control (MAC) and
Physical Layer (PHY) port. This driver provides support for MAC and
DMA-specific initialization and runtime TX/RX operations.

Key features:
- MSI-X interrupts for TX/RX DMA channels.
- Multiple TX/RX DMA channel support with exclusive bottom-half.
- Implements a circular descriptor ring architechture with
  producer-consumer semantics for high performance pkt processing.
- Full duplex support for 10/100/1000 Mbps.
- Half duplex support for 10/100 Mbps.
- Auto-negotiation for 10/100/1000 Mbps.
- MTU customization for flexible packet sizes.
- MAC address filtering based on:
  - Random MAC generation.
  - Local-mac-address mentioned in device tree.
  - EEPROM pre-programmed mac address.
  - Setting mac address via net shell.
- Support for multiple Ethernet interface instances.

Signed-off-by: Vijayakannan Ayyathurai <vijayakannan.ayyathurai@intel.com>
Signed-off-by: Ling Pei Lee <pei.lee.ling@intel.com>

Vijayakannan committed:
dts: x86: Add i226 device-tree config for Intel Alderlake

Foxville LM (0x125B) i226 variant and Intel Alder Lake platform was
used for developing and stabilizing the i226 Ethernet device driver.
However, users can reuse the provided device tree models as a reference
when enabling the support for other i226 variants and platforms.

This device-tree model include essential configurations for the i226
Ethernet controller, such as PCIe settings, interrupt mappings, Phy
MDIO, and DMA descriptor configurations.

Signed-off-by: Vijayakannan Ayyathurai <vijayakannan.ayyathurai@intel.com>

Vijayakannan committed:
samples: net: zperf: Enable i226 config with Intel Alderlake

The Zperf sample application was chosen to demonstrate basic network
functionality and high-performance use cases. This application serves
as a reference for users who wish to enable other network sample
applications.

In addition to the essential configurations for the Intel i226 Ethernet
controller, stack-specific configurations were added to ensure stability
under heavy network loads. These configurations include adjustments to
buffer sizes, interrupt handling, and DMA descriptor management.

Signed-off-by: Vijayakannan Ayyathurai <vijayakannan.ayyathurai@intel.com>

Vijayakannan commented:
Hi @maass-hamburg,

All the review comments were addressed in this PR version 7. Please review and share feedback. Thanks, Vijay

Changes since v6:

	- Use generic phy driver for i226. (Test Coverage: Got Satisfactory performance results)
		○ Replace the compatible string in device tree to use "ethernet-phy" generic driver.
		○ Remove drivers/ethernet/phy/phy_intel_igc.c
		○ Remove dts/bindings/ethernet/phy/intel,igc-phy.yaml
		○ Clean up drivers/ethernet/phy/CMakeLists.txt.
		○ Clean up drivers/ethernet/phy/Kconfig.

	- Remove Miscellaneous interrupts handling codes from the mac driver.
		○ eth_intel_igc_misc_isr
		○ eth_intel_igc_connect_misc_msix_vector
		○ Clean eth_intel_igc_pcie_msix_setup
		○ Clean eth_intel_igc_map_intr_to_vector
		○ eth_intel_igc_intr_enable

	- Remove INTEL_IGC_MAC_ASSIGN_ADDR unused macro.
	- Move interrupt-parent to parent platform driver.
	- Clean up the PR description.

Vijayakannan commented on drivers/ethernet/intel/eth_intel_igc.c:
```diff
@@ -0,0 +1,1471 @@
+		return;
+	}
+
+	eth_intel_igc_intr_enable(dev);
```
Now i226 uses the generic ethernet-phy driver. I believe PR version 7 resolves this discussion. Please review once and approve if there is no more feedback.

maass-hamburg commented on drivers/ethernet/intel/eth_intel_igc.c:
```diff
@@ -0,0 +1,1363 @@
+#define NUM_RX_DESC(n) DT_INST_PROP(n, num_rx_desc)
+#define NUM_MSIX(n)    NUM_QUEUES(n)
+
+#define INTEL_IGC_MAC_CONFIG_IRQ(n)                                                                \
+	static void eth##n##_cfg_irq(const struct device *dev)                                     \
+	{                                                                                          \
+		eth_intel_igc_map_intr_to_vector(dev);                                             \
+		eth_intel_igc_init_queue_work_##n(dev);                                            \
+		eth_intel_igc_init_speed(dev->data);                                               \
+	}
```
eth_intel_igc_map_intr_to_vector(dev);
and
eth_intel_igc_init_speed(dev->data);
should be called outside this instance specific function

maass-hamburg commented on drivers/ethernet/intel/eth_intel_igc.c:
```diff
@@ -0,0 +1,1363 @@
+
+	eth_intel_igc_set_mac_filter(dev, DEST_ADDR, data->mac_addr, 0, 0);
+
+	/* Configure interrupt vectors and handlers */
+	cfg->config_func(dev);
```
should happen in normal init

maass-hamburg commented on drivers/ethernet/intel/eth_intel_igc.c:
```diff
@@ -0,0 +1,1363 @@
+	uint32_t mac_addr_lo =
+		(mac_addr[3] << 24) | (mac_addr[2] << 16) | (mac_addr[1] << 8) | (mac_addr[0] << 0);
+
+	igc_write(base + INTEL_IGC_RAH(index), mac_addr_hi);
+	igc_reg_refresh(base);
+	igc_write(base + INTEL_IGC_RAL(index), mac_addr_lo);
```
what are these registers exactly? because you also read from them in `eth_intel_igc_get_nvm_mac_addr()`? Do you have to set these registers, for it to work?

Vijayakannan replied:
> what are these registers exactly?
> Do you have to set these registers, for it to work?
RAH and RAL registers form the MAC address filter table in the i226 hardware. These registers are essential for the initial packet filtering stage - the hardware uses them to determine which packets to accept based on their destination MAC address. If these registers are not properly configured, packet reception will fail.

> because you also read from them in `eth_intel_igc_get_nvm_mac_addr()`? 

According to the i226 design specification, the RAH and RAL registers are automatically loaded with a valid MAC address when the hardware is configured to retrieve the MAC address from EEPROM. However, during our initial driver development, we discovered that these valid values are only available after a 2-microsecond delay following MAC reset.

The WAIT_FOR macro ensures we don't read the registers before the EEPROM values are properly loaded. Without this delay, data->mac_addr would be assigned invalid reset values, leading to subsequent TX/RX failures.

maass-hamburg commented on drivers/ethernet/intel/eth_intel_igc.c:
```diff
@@ -0,0 +1,1363 @@
+
+#define DT_DRV_COMPAT intel_igc_mac
+
+static uint32_t igc_read(mm_reg_t reg_addr, uint32_t offset)
+{
+	return sys_read32(reg_addr + offset);
+}
+
+static void igc_write(mm_reg_t reg_addr, uint32_t val)
+{
+	sys_write32(val, reg_addr);
+}
```
these functions don't seem to to provide any real extra value, instead of directly using `sys_read32` and `sys_write32`. Why not using them directly?

maass-hamburg commented on drivers/ethernet/intel/eth_intel_igc.c:
```diff
@@ -0,0 +1,1349 @@
+ * @brief Generate TX and RX interrupt handling functions as per queue.
+ */
+#define INTEL_IGC_SETUP_QUEUE_WORK_EXP(n, queue)                                                   \
+	static void eth_tx_irq_queue##n##_##queue(struct k_work *item)                             \
+	{                                                                                          \
+		struct k_work *work = item;                                                        \
```
```suggestion
	static void eth_tx_irq_queue##n##_##queue(struct k_work *work)                             \
	{                                                                                          \
```

maass-hamburg commented on drivers/ethernet/intel/eth_intel_igc.c:
```diff
@@ -0,0 +1,1349 @@
+ * @brief Initializes the runtime data structures of each driver instance.
+ */
+#define INTEL_IGC_MAC_DATA(n)                                                                      \
+	static struct eth_intel_igc_mac_tx tx_##n = {                                              \
+		.sem = tx_ring_lock_##n,                                                           \
+		.ring_wr_ptr = tx_ring_wr_ptr_##n,                                                 \
+		.ring_rd_ptr = tx_ring_rd_ptr_##n,                                                 \
+		.pkt = (struct net_pkt **)tx_pkt_##n,                                              \
+		.frag = (struct net_buf **)tx_frag_##n,                                            \
+	};                                                                                         \
+	static struct eth_intel_igc_mac_rx rx_##n = {                                              \
+		.sem = rx_ring_lock_##n,                                                           \
+		.ring_wr_ptr = rx_ring_wr_ptr_##n,                                                 \
+		.ring_rd_ptr = rx_ring_rd_ptr_##n,                                                 \
+	};                                                                                         \
+	static struct eth_intel_igc_mac_data eth_data_##n = {                                      \
+		.tx = &tx_##n,                                                                     \
+		.rx = &rx_##n,                                                                     \
```
put struct eth_intel_igc_mac_tx and struct eth_intel_igc_mac_rx directly in struct eth_intel_igc_mac_data, not just as a pointer

maass-hamburg commented on drivers/ethernet/intel/eth_intel_igc.c:
```diff
@@ -0,0 +1,1349 @@
+	static void eth##n##_cfg_irq(const struct device *dev);                                    \
+	static const struct eth_intel_igc_mac_cfg eth_cfg_##n = {                                  \
+		.platform = DEVICE_DT_GET(DT_INST_PARENT(n)),                                      \
+		.phy = DEVICE_DT_GET(DT_NODELABEL(ethphy##n)),                                     \
```
phy should come from the `phy-handle` phandle

Vijayakannan commented on drivers/ethernet/intel/eth_intel_igc.c:
```diff
@@ -0,0 +1,1349 @@
+	/* Turn off VLAN filters */
+	config &= ~INTEL_IGC_RCTL_VFE;
+	config |= INTEL_IGC_RCTL_EN | INTEL_IGC_RCTL_BAM |
+		  /* Enable long packet receive */
+		  INTEL_IGC_RCTL_LPE |
```
> What is long packet receive? Zephyr doesn't support jumbo frames btw.

I will clean it up, because 1522 is the current long packet requirement.

Vijayakannan commented:
Hi @maass-hamburg, please review PR revision v9.

Changes since PR version v8:
- Remove 0 left and right shift usages in code.
- Cleanup eth_intel_igc_get_sa_family
- Remove Long packet receive support. (test the speed with this support)
- Cleanup INTEL_IGC_SETUP_QUEUE_WORK_EXP macro
- Cleanup INTEL_IGC_MAC_DATA
- Use phy-handle to fetch ethernet-phy node in dts.

Vijayakannan commented:
Hi @maass-hamburg, could you please share your feedback on this updated PR version?

Vijayakannan commented on drivers/ethernet/intel/eth_intel_igc_priv.h:
```diff
@@ -0,0 +1,500 @@
+struct eth_intel_igc_mac_data {
+	struct net_if *iface;
+	const struct device *mac;
+	struct eth_intel_igc_mac_tx *tx;
+	struct eth_intel_igc_mac_rx *rx;
```
> I meant this, when I said to put it in it.

Got it. This change requires updating the overall code. I can make the necessary modifications and update this PR.

Please let me know if there are any other review comments to address in the next version. I need to expedite this PR due to the long pending pull request.
